src/hotspot/share/opto/mulnode.cpp

```diff
@@ -670,9 +670,13 @@ const Type *AndINode::mul_ring( const Type *t0, const Type *t1 ) const {
   return and_value<TypeInt>(r0, r1);
 }
 
+// Is expr a neutral element wrt addition under mask?
+static bool AndIL_is_zero_element(const PhaseGVN* phase, const Node* expr, const Node* mask, BasicType bt);
+
 const Type* AndINode::Value(PhaseGVN* phase) const {
   // patterns similar to (v << 2) & 3
-  if (AndIL_is_always_zero(phase, in(1), in(2), T_INT, true)) {
+  if (AndIL_is_zero_element(phase, in(1), in(2), T_INT) ||
+      AndIL_is_zero_element(phase, in(2), in(1), T_INT)) {
```

Comment:
I find it easier to reason about the "reverse" check when we simply expand it here.